crates/sp-lightclient/src/lib.rs

```diff
@@ -374,6 +402,92 @@ impl<Header: HeaderT, Store: Storage<Header>> HeaderImporter<Header, Store> {
         Ok(())
     }
 
+    /// Derives next salt derivation info with respect to parent salt derivation info.
+    fn next_salt_derivation_info(
+        &self,
+        header: &Header,
+        parent_salt_derivation_info: &SaltDerivationInfo,
+    ) -> Result<SaltDerivationInfo, ImportError<Header>> {
+        let constants = self.store.chain_constants();
+        let eon_duration = constants.eon_duration;
+        let genesis_slot = self.find_genesis_slot(header)?;
+        let pre_digest = extract_pre_digest(header)?;
+
+        // if the salt is not revealed yet, check if salt will be revealed at this header
+        let maybe_randomness = if parent_salt_derivation_info.maybe_randomness.is_none() {
+            let next_salt_reveal_slot = constants
+                .next_salt_reveal_interval
+                .checked_add(
+                    parent_salt_derivation_info
+                        .eon_index
+                        .checked_mul(eon_duration)
+                        .and_then(|res| res.checked_add(u64::from(genesis_slot)))
+                        .expect("eon start slot should fit into u64"),
+                )
+                .ok_or(ImportError::ArithmeticError(ArithmeticError::Overflow))?;
+
+            // salt will be revealed after importing this header.
+            // derive randomness at this header and store it for later verification
+            if pre_digest.slot >= next_salt_reveal_slot {
+                let randomness = derive_randomness(
+                    &Into::<PublicKey>::into(&pre_digest.solution.public_key),
+                    pre_digest.solution.tag,
+                    &pre_digest.solution.tag_signature,
+                )
+                .map_err(|_err| {
+                    ImportError::DigestError(DigestError::NextDigestDerivationError(
+                        ErrorDigestType::NextSalt,
+                    ))
+                })?;
+                Some(randomness)
+            } else {
+                None
+            }
+        } else {
+            // salt is already revealed
+            parent_salt_derivation_info.maybe_randomness
+        };
+
+        // check if the eon is about to be changed
+        let maybe_next_index = subspace_verification::derive_next_eon_index(
+            parent_salt_derivation_info.eon_index,
+            eon_duration,
+            u64::from(genesis_slot),
+            u64::from(pre_digest.slot),
+        );
+
+        // if eon has changed, reset randomness
+        if let Some(next_eon_index) = maybe_next_index {
```

Comment:
This can happen at the same time as `maybe_randomness` re-generation. Which if I read this correctly will not actually result in updated salt this time. I think it would be worthwhile adding such case to the test cases and run light client logic against actual history produced by Subspace in addition to just manually composed test cases, there might be other edge cases I can't compile in my head.

Reply:
nvm. You are correct. It is possible :-D

Reply:
> run light client logic against actual history produced by Subspace in addition to just manually composed test cases, there might be other edge cases I can't compile in my head.

Yes this is the plan as I'm blanking on thinking potential edge cases for tests. My plan was to bring in the networking and connect it to multiple dev and testnets for potential issues we might not have covered